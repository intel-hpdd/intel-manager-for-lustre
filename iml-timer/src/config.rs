// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::error::TimerError;
use futures::TryFutureExt;
use serde::{Deserialize, Serialize};
use tokio::{
    fs::{remove_file, File},
    prelude::*,
};

#[derive(Debug, Deserialize, Serialize)]
pub struct ConfigDetails {
    config_id: String,
    file_prefix: String,
    timer_config: String,
    service_config: String,
}

#[derive(Deserialize, Serialize)]
pub struct DeleteConfigDetails {
    config_id: String,
}

pub struct ConfigFile {
    pub name: String,
    pub content: String,
}

pub struct ConfigFiles {
    pub timer_file: ConfigFile,
    pub service_file: ConfigFile,
}

pub fn unit_name(file_prefix: &str, fid: &str) -> String {
    format!("{}-{}", file_prefix, fid)
}

pub fn timer_file(file_prefix: &str, fid: &str) -> String {
    format!("/etc/systemd/system/{}.timer", unit_name(file_prefix, fid))
}

pub fn service_file(file_prefix: &str, fid: &str) -> String {
    format!(
        "/etc/systemd/system/{}.service",
        unit_name(file_prefix, fid)
    )
}

pub fn get_config(config: ConfigDetails) -> (String, String, ConfigFiles) {
    iml_tracing::tracing::debug!("config: {:?}", config);
    (
        config.file_prefix.clone(),
        config.config_id.clone(),
        ConfigFiles {
            timer_file: ConfigFile {
                name: timer_file(&config.file_prefix, config.config_id.as_str()),
                content: config.timer_config,
            },
            service_file: ConfigFile {
                name: service_file(&config.file_prefix, config.config_id.as_str()),
                content: config.service_config,
            },
        },
    )
}

pub async fn write_config_to_file(buf: &[u8], file: &str) -> tokio::io::Result<()> {
    let mut file = File::create(file).await?;
    file.write_all(buf).await?;
    Ok(())
}

pub async fn write_config_files(configs: ConfigFiles) -> tokio::io::Result<()> {
    write_config_to_file(
        configs.timer_file.content.as_bytes(),
        &configs.timer_file.name,
    )
    .await?;

    write_config_to_file(
        configs.service_file.content.as_bytes(),
        &configs.service_file.name,
    )
    .await?;

    Ok(())
}

pub async fn write_configs(
    (file_prefix, config_id, configs): (String, String, ConfigFiles),
) -> Result<(String, String), warp::Rejection> {
    write_config_files(configs)
        .map_err(TimerError::IoError)
        .map_err(warp::reject::custom)
        .await?;

    Ok((file_prefix, config_id))
}

pub async fn delete_config(
    config: &str,
    file_prefix: &str,
    config_id: &str,
) -> Result<(String, String), warp::Rejection> {
    remove_file(config)
        .map_err(TimerError::IoError)
        .map_err(warp::reject::custom)
        .await?;

    Ok((file_prefix.to_string(), config_id.to_string()))
}
