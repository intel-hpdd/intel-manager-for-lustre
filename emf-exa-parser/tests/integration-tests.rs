// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use emf_exa_parser::config::ExascalerConfiguration;
use insta::{assert_json_snapshot, Settings};

#[test]
fn parse_exascaler_conf_from_file_test() {
    let json = include_str!("fixtures/exascaler.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}

#[test]
fn parse_exascaler_conf_from_file_test_ai400x() {
    let json = include_str!("fixtures/exascaler_ai400x.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}

#[test]
fn parse_exascaler_conf_from_file_test_zpool() {
    let json = include_str!("fixtures/exascaler_zpool.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}

#[test]
fn parse_exascaler_conf_from_file_test_bonded() {
    let json = include_str!("fixtures/exascaler_es7k_with_extMDS.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}
