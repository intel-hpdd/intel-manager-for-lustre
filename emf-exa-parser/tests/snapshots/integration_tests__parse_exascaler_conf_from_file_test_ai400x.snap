---
source: emf-exa-parser/tests/integration-tests.rs
expression: exa

---
{
  "ha_settings": {
    "corosync_nics": [
      "ring0",
      "ring1"
    ],
    "crypto_cipher": "aes256",
    "crypto_hash": "sha1",
    "dampen_ifspeed": 20,
    "dampen_ping": 20,
    "failover_policy": "standard",
    "ha_group_count": 1,
    "ha_groups": [
      [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ]
    ],
    "lustre_start_timeout": 450,
    "max_messages": 17,
    "mcastport": 5405,
    "netmtu": 1500,
    "no_quorum_policy": "freeze",
    "rrp_mode": "passive",
    "secauth": "on",
    "stonith_timeout": 60,
    "start_on_boot": true,
    "transport": "unicast",
    "type": "corosync",
    "window_size": 50,
    "zpool_monitor_timeout": 60
  },
  "global_settings": {
    "clients_list": [],
    "cluster_name": null,
    "conf_param_tunings": null,
    "email_list": [],
    "email_domain": null,
    "email_relay": null,
    "extra_hosts_start": [],
    "extra_hosts_end": [],
    "fs_list": [
      "ai400hyb"
    ],
    "host_list": [
      "ai400x-7f94-vm00",
      "ai400x-7f94-vm01",
      "ai400x-7f94-vm02",
      "ai400x-7f94-vm03"
    ],
    "hsm_active": false,
    "kdump_path": "/scratch/crash",
    "lnet_mr_fault_sensitive": false,
    "log_dir": "/scratch/log",
    "mdt_backup": "none",
    "mdt_backup_dir": "/scratch",
    "mgs_fs": "ai400hyb",
    "ntp_list": [],
    "password_policy": "plain-text",
    "pingd": true,
    "s2a_list": [],
    "s2a_pass": "password",
    "s2a_user": "admin",
    "set_param_tunings": {
      "obdfilter.*.brw_size": "16",
      "osd-ldiskfs.*.read_cache_enable": "0",
      "osd-ldiskfs.*.writethrough_cache_enable": "0"
    },
    "sfa_list": [
      "AI400X-7f94"
    ],
    "shadow_conf": false,
    "timezone": null,
    "vg_activation_list": [
      "auto"
    ]
  },
  "pool_settings": {},
  "sfa_settings": {
    "AI400X-7f94": {
      "controllers": [
        "10.0.3.1",
        "10.0.3.4"
      ],
      "name": "AI400X-7f94",
      "password": "user",
      "user": "user"
    }
  },
  "zpool_settings": null,
  "host_defaults_settings": {
    "base_ip": {},
    "bonding_mode": null,
    "grub_args": "spectre_v2=off nopti",
    "host_sfa_list": [
      "AI400X-7f94"
    ],
    "ipmi_delay": null,
    "ipmi_method": null,
    "ipmi_monitor": null,
    "ipmi_power_wait": 5,
    "lnets": [
      "o2ib0(ib0,ib1)"
    ],
    "modprobe_cfg": null,
    "rest_cert_ca": null,
    "rest_cert_crl": null,
    "rest_cert_server": null,
    "rest_cert_server_key": null,
    "rest_ext_nic": null,
    "rest_int_nic": null,
    "rest_keepalived_nic": null,
    "rest_primary_nic": null,
    "nic_list": [
      "mgmt0",
      "mgmt2",
      "ib0",
      "ib1"
    ],
    "nics": {
      "ib0": {
        "cfg": null,
        "device": "ib0",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": "255.255.0.0",
        "slaves": null
      },
      "ib1": {
        "cfg": null,
        "device": "ib1",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": "255.255.0.0",
        "slaves": null
      },
      "mgmt0": {
        "cfg": null,
        "device": "mgmt0",
        "gateway": "10.36.0.1",
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": "255.255.0.0",
        "slaves": null
      },
      "mgmt2": {
        "cfg": null,
        "device": "mgmt2",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": "255.255.255.0",
        "slaves": null
      }
    },
    "ring0": "mgmt2",
    "ring1": "ib0",
    "serial_port": "ttyS0",
    "serial_speed": "115200",
    "stonith_pass": null,
    "stonith_user": null,
    "stonith_type": "sfa_vm"
  },
  "hosts_settings": {
    "ai400x-7f94-vm00": {
      "fs_list": [
        "ai400hyb"
      ],
      "bonding_mode": null,
      "grub_args": "spectre_v2=off nopti",
      "ha_group": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "AI400X-7f94"
      ],
      "ipmi_delay": null,
      "ipmi_method": null,
      "ipmi_monitor": null,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "o2ib0": [
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ],
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ]
        ]
      },
      "lnet_nics": [
        "ib0",
        "ib1"
      ],
      "lnets": [
        "o2ib0(ib0,ib1)"
      ],
      "mdt_base_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "mdt_list": {
        "ai400hyb": [
          0
        ]
      },
      "modprobe_cfg": null,
      "name": "ai400x-7f94-vm00",
      "nic_list": [
        "mgmt0",
        "mgmt2",
        "ib0",
        "ib1"
      ],
      "nics": {
        "ib0": {
          "cfg": null,
          "device": "ib0",
          "gateway": null,
          "ip": "172.16.254.194",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "ib1": {
          "cfg": null,
          "device": "ib1",
          "gateway": null,
          "ip": "172.16.255.194",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt0": {
          "cfg": null,
          "device": "mgmt0",
          "gateway": "10.36.0.1",
          "ip": "10.36.5.194",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.36.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt2": {
          "cfg": null,
          "device": "mgmt2",
          "gateway": null,
          "ip": "10.0.3.2",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.0.3.0",
          "netmask": "255.255.255.0",
          "slaves": null
        }
      },
      "peers": [
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "oid": "0x10000000",
      "ost_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "ost_list": {
        "ai400hyb": [
          0,
          1,
          8
        ]
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "mgmt2",
      "ring1": "ib0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "user",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "sfa_vm",
      "stonith_user": "user",
      "sysctl": {
        "net.ipv4.conf.all.arp_announce": "2",
        "net.ipv4.conf.all.arp_ignore": "1",
        "net.ipv4.conf.all.rp_filter": "0",
        "net.ipv4.conf.defalut.arp_announce": "2",
        "net.ipv4.conf.default.arp_ignore": "1",
        "net.ipv4.conf.default.rp_filter": "2",
        "vm.min_free_kbytes": "2097152"
      }
    },
    "ai400x-7f94-vm01": {
      "fs_list": [
        "ai400hyb"
      ],
      "bonding_mode": null,
      "grub_args": "spectre_v2=off nopti",
      "ha_group": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "AI400X-7f94"
      ],
      "ipmi_delay": null,
      "ipmi_method": null,
      "ipmi_monitor": null,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "o2ib0": [
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ],
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ]
        ]
      },
      "lnet_nics": [
        "ib0",
        "ib1"
      ],
      "lnets": [
        "o2ib0(ib0,ib1)"
      ],
      "mdt_base_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "mdt_list": {
        "ai400hyb": [
          1
        ]
      },
      "modprobe_cfg": null,
      "name": "ai400x-7f94-vm01",
      "nic_list": [
        "mgmt0",
        "mgmt2",
        "ib0",
        "ib1"
      ],
      "nics": {
        "ib0": {
          "cfg": null,
          "device": "ib0",
          "gateway": null,
          "ip": "172.16.254.195",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "ib1": {
          "cfg": null,
          "device": "ib1",
          "gateway": null,
          "ip": "172.16.255.195",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt0": {
          "cfg": null,
          "device": "mgmt0",
          "gateway": "10.36.0.1",
          "ip": "10.36.5.195",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.36.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt2": {
          "cfg": null,
          "device": "mgmt2",
          "gateway": null,
          "ip": "10.0.3.5",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.0.3.0",
          "netmask": "255.255.255.0",
          "slaves": null
        }
      },
      "peers": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm03",
        "ai400x-7f94-vm02"
      ],
      "oid": "0x10008000",
      "ost_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "ost_list": {
        "ai400hyb": [
          2,
          3,
          9
        ]
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "mgmt2",
      "ring1": "ib0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "user",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "sfa_vm",
      "stonith_user": "user",
      "sysctl": {
        "net.ipv4.conf.all.arp_announce": "2",
        "net.ipv4.conf.all.arp_ignore": "1",
        "net.ipv4.conf.all.rp_filter": "0",
        "net.ipv4.conf.defalut.arp_announce": "2",
        "net.ipv4.conf.default.arp_ignore": "1",
        "net.ipv4.conf.default.rp_filter": "2",
        "vm.min_free_kbytes": "2097152"
      }
    },
    "ai400x-7f94-vm02": {
      "fs_list": [
        "ai400hyb"
      ],
      "bonding_mode": null,
      "grub_args": "spectre_v2=off nopti",
      "ha_group": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "AI400X-7f94"
      ],
      "ipmi_delay": null,
      "ipmi_method": null,
      "ipmi_monitor": null,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "o2ib0": [
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ],
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ]
        ]
      },
      "lnet_nics": [
        "ib0",
        "ib1"
      ],
      "lnets": [
        "o2ib0(ib0,ib1)"
      ],
      "mdt_base_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "mdt_list": {
        "ai400hyb": [
          2
        ]
      },
      "modprobe_cfg": null,
      "name": "ai400x-7f94-vm02",
      "nic_list": [
        "mgmt0",
        "mgmt2",
        "ib0",
        "ib1"
      ],
      "nics": {
        "ib0": {
          "cfg": null,
          "device": "ib0",
          "gateway": null,
          "ip": "172.16.254.196",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "ib1": {
          "cfg": null,
          "device": "ib1",
          "gateway": null,
          "ip": "172.16.255.196",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt0": {
          "cfg": null,
          "device": "mgmt0",
          "gateway": "10.36.0.1",
          "ip": "10.36.5.196",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.36.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt2": {
          "cfg": null,
          "device": "mgmt2",
          "gateway": null,
          "ip": "10.0.3.3",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.0.3.0",
          "netmask": "255.255.255.0",
          "slaves": null
        }
      },
      "peers": [
        "ai400x-7f94-vm03",
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01"
      ],
      "oid": "0x10000001",
      "ost_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "ost_list": {
        "ai400hyb": [
          4,
          5,
          10
        ]
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "mgmt2",
      "ring1": "ib0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "user",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "sfa_vm",
      "stonith_user": "user",
      "sysctl": {
        "net.ipv4.conf.all.arp_announce": "2",
        "net.ipv4.conf.all.arp_ignore": "1",
        "net.ipv4.conf.all.rp_filter": "0",
        "net.ipv4.conf.defalut.arp_announce": "2",
        "net.ipv4.conf.default.arp_ignore": "1",
        "net.ipv4.conf.default.rp_filter": "2",
        "vm.min_free_kbytes": "2097152"
      }
    },
    "ai400x-7f94-vm03": {
      "fs_list": [
        "ai400hyb"
      ],
      "bonding_mode": null,
      "grub_args": "spectre_v2=off nopti",
      "ha_group": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "AI400X-7f94"
      ],
      "ipmi_delay": null,
      "ipmi_method": null,
      "ipmi_monitor": null,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "o2ib0": [
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ],
          [
            "172.16.254.194",
            "172.16.255.194"
          ],
          [
            "172.16.254.195",
            "172.16.255.195"
          ],
          [
            "172.16.254.196",
            "172.16.255.196"
          ],
          [
            "172.16.254.197",
            "172.16.255.197"
          ]
        ]
      },
      "lnet_nics": [
        "ib0",
        "ib1"
      ],
      "lnets": [
        "o2ib0(ib0,ib1)"
      ],
      "mdt_base_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "mdt_list": {
        "ai400hyb": [
          3
        ]
      },
      "modprobe_cfg": null,
      "name": "ai400x-7f94-vm03",
      "nic_list": [
        "mgmt0",
        "mgmt2",
        "ib0",
        "ib1"
      ],
      "nics": {
        "ib0": {
          "cfg": null,
          "device": "ib0",
          "gateway": null,
          "ip": "172.16.254.197",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "ib1": {
          "cfg": null,
          "device": "ib1",
          "gateway": null,
          "ip": "172.16.255.197",
          "is_bonded": false,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt0": {
          "cfg": null,
          "device": "mgmt0",
          "gateway": "10.36.0.1",
          "ip": "10.36.5.197",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.36.0.0",
          "netmask": "255.255.0.0",
          "slaves": null
        },
        "mgmt2": {
          "cfg": null,
          "device": "mgmt2",
          "gateway": null,
          "ip": "10.0.3.6",
          "is_bonded": false,
          "master": null,
          "netaddr": "10.0.3.0",
          "netmask": "255.255.255.0",
          "slaves": null
        }
      },
      "peers": [
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm00"
      ],
      "oid": "0x10008001",
      "ost_device_paths": {
        "ai400hyb": "/dev/ddn"
      },
      "ost_list": {
        "ai400hyb": [
          6,
          7,
          11
        ]
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "mgmt2",
      "ring1": "ib0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "user",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "sfa_vm",
      "stonith_user": "user",
      "sysctl": {
        "net.ipv4.conf.all.arp_announce": "2",
        "net.ipv4.conf.all.arp_ignore": "1",
        "net.ipv4.conf.all.rp_filter": "0",
        "net.ipv4.conf.defalut.arp_announce": "2",
        "net.ipv4.conf.default.arp_ignore": "1",
        "net.ipv4.conf.default.rp_filter": "2",
        "vm.min_free_kbytes": "2097152"
      }
    }
  },
  "rest_settings": null,
  "emf_settings": {
    "ip": "10.36.6.194",
    "cidr": 16,
    "size": "80G",
    "enabled": true,
    "nic": null
  },
  "fs_settings": {
    "ai400hyb": {
      "backfs": "ldiskfs",
      "cur_mdt_index": 4,
      "cur_ost_index": 0,
      "default_ost_count": 3,
      "default_mdt_count": 1,
      "dom_enabled": false,
      "dom_max_file_size": 65536,
      "host_list": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "hsm_active": false,
      "log_dir": "/scratch/log/ai400hyb",
      "mds_list": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "mdt_base_device_path": "/dev/ddn",
      "mdt_failback": true,
      "mdt_list": {
        "ai400x-7f94-vm00": [
          0
        ],
        "ai400x-7f94-vm01": [
          1
        ],
        "ai400x-7f94-vm02": [
          2
        ],
        "ai400x-7f94-vm03": [
          3
        ]
      },
      "mdt_mke2fs_opts": [
        "-m1",
        "-E",
        "lazy_itable_init=0,lazy_journal_init=0,nodiscard",
        "-O",
        "project"
      ],
      "mdt_mount_opts": "max_sectors_kb=0",
      "mgs_device_path": null,
      "mgs_list": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01"
      ],
      "mgt_mount_opts": "max_sectors_kb=0",
      "mdt_opts": "--param mdt.identity_upcall=NONE",
      "mdt_parts": null,
      "mdt_size": "90%",
      "mgs_failback": true,
      "mgs_internal": true,
      "mgs_size": "2G",
      "mgs_standalone": true,
      "mmp_update_interval": 5,
      "name": "ai400hyb",
      "oss_list": [
        "ai400x-7f94-vm00",
        "ai400x-7f94-vm01",
        "ai400x-7f94-vm02",
        "ai400x-7f94-vm03"
      ],
      "ost_device_path": "/dev/ddn",
      "ost_list": {
        "ai400x-7f94-vm00": [
          0,
          1,
          8
        ],
        "ai400x-7f94-vm01": [
          2,
          3,
          9
        ],
        "ai400x-7f94-vm02": [
          4,
          5,
          10
        ],
        "ai400x-7f94-vm03": [
          6,
          7,
          11
        ]
      },
      "ost_lnet_list": {},
      "ost_mke2fs_opts": [
        "-m1",
        "-i",
        "131072",
        "-E",
        "lazy_itable_init=0,lazy_journal_init=0,nodiscard",
        "-O",
        "project"
      ],
      "ost_mount_opts": "max_sectors_kb=0",
      "ost_opts": null,
      "pools": [],
      "total_mdt_count": 4,
      "total_ost_count": 12,
      "tune_ost_mke2fs_opts": [
        "for size > 140737488355328 apply '-E lazy_itable_init=1,nodiscard -N 1073741824 -O project'"
      ]
    }
  }
}
